core/src/main/java/org/apache/brooklyn/enricher/stock/reducer/Reducer.java

```diff
@@ -0,0 +1,187 @@
+            }
+        }
+        
+        String reducerName = config().get(REDUCER_FUNCTION_TRANSFORMATION);
+        Function<List<S>, T> reducerFunction = (Function) config().get(REDUCER_FUNCTION);
+        if(reducerFunction == null){
+            Map<String, ?> parameters = config().get(PARAMETERS);
+            reducerFunction = createReducerFunction(reducerName, parameters);
+        }
+
+        this.reducerFunction = reducerFunction;
+        Preconditions.checkState(sensorListTemp.size() > 0, "Nothing to reduce");
+
+        for (Sensor<S> sensor : sensorListTemp) {
+            subscribe(producer, sensor, this);
+        }
+
+        subscribedSensors = ImmutableList.copyOf(sensorListTemp);
+    }
+
+    protected abstract Function<List<S>, T> createReducerFunction(String reducerName, Map<String, ?> parameters);
+    
+    @SuppressWarnings("unchecked")
+    @Override
+    public void onEvent(SensorEvent<Object> event) {
+        Sensor<T> destinationSensor = (Sensor<T>) getConfig(TARGET_SENSOR);
+
+        List<S> values = Lists.newArrayList();
+
+        for (AttributeSensor<S> sourceSensor : subscribedSensors) {
+            S resolvedSensorValue = entity.sensors().get(sourceSensor);
+            if (resolvedSensorValue == null) {
+                // only apply function if all values are resolved
+                return;
+            }
+
+            values.add(resolvedSensorValue);
+        }
+
+        Object result = reducerFunction.apply(values);
+
+        if (LOG.isTraceEnabled()) LOG.trace("enricher {} got {}, propagating via {} as {}",
+                new Object[] {this, event, entity, reducerFunction, destinationSensor});
+
+        emit((Sensor<T>)destinationSensor, result);
+    }
+   
+    public static class JoinerReducerFunction<A> implements Function<List<A>, String> {
+        
+        private Object separator;
+
+        public JoinerReducerFunction(Object separator) {
+            this.separator = (separator == null) ? ", " : separator;
+        }
+
+        @Override
+        public String apply(List<A> input) {
+            
+            StringBuilder sb = new StringBuilder();
+            Iterator<A> it = input.iterator();
+            while(it.hasNext()) {
+                sb.append(it.next().toString());
+                if(it.hasNext()){
+                    sb.append(separator);
+                }
+            }
+            return sb.toString();
+        }
+        
+    }
+
+    public static class JoinerFunction extends JoinerReducerFunction<String>{
+        
+        public JoinerFunction(Object separator) {
+            super(separator);
+        }
+    }
+
+    public static class ToStringReducerFunction<A> implements Function<List<A>, String> {
+
+        @Override
+        public String apply(List<A> input) {
+            return input.toString();
+        }
+        
+    }
+    
+    public static class FormatStringReducerFunction<T> implements Function<List<T>, String> {
```

Comment:
See `StringFunctions.formatterForArray` (might need a new variant for `Iterable<T>`).